README.md
# Advent of Code in Haskell

```
cd src
stack runhaskell Day1.hs

```


ChangeLog.md
# Changelog for adventcode

## Unreleased changes


src/Day2.hs

import Data.Function
import Data.List
import Data.Maybe (fromMaybe)
import Data.List.Split (splitOn)
import System.IO

main:: IO ()
main = do  
    handle <- openFile "advent2.txt" ReadMode  
    contents <- hGetContents handle  
    -- print contents  
    -- print $ result [[5, 1, 9, 5], [7, 5, 3], [2, 4, 6]]
    print $ result $ convert contents
    hClose handle  

convert :: String -> [[Int]]
convert = 
    fmap (\list -> fmap (fromMaybe 0 . readMaybe) list) . fmap (splitOn "\t") . splitOn "\n"


result :: [[Int]] -> Int
result input = 
    input
    & fmap checkSum
    & sum


checkSum :: [Int] -> Int
checkSum l =
    let
        sortList = sort l
    in
        last sortList - head sortList


readMaybe :: (Read a) => String -> Maybe a
readMaybe s = case reads s of
                [(x, "")] -> Just x
                _ -> Nothing

src/Day6.hs
import Data.List
import Data.Maybe

main = do 
    -- print $ findCycle [0, 2, 7, 0]
    print $ spread $ setZero $ findHighest [0, 2, 7, 0]
    print $ setZero $ findHighest [0, 2, 7, 0]
    print $ findCycle [0, 2, 7, 0]
    print $ findCycle [4,1,15,12,0,9,9,5,5,8,7,3,14,5,12,3]
    print $ spread (7,2,[0,2,0,0])
    -- print $ setZero 7 [0, 2, 7, 0]

    
findCycle :: [Int] -> Int
findCycle list = 
    findCycle' 0 list []
    where
        findCycle' :: Int -> [Int] -> [[Int]] -> Int
        findCycle' cycle list seen = 
            case list `elem` seen of
                True -> 
                    cycle
                False ->
                    findCycle' (cycle + 1) (redistribute list) (list:seen)

redistribute :: [Int] -> [Int]        
redistribute list =
    spread $ setZero $ findHighest list

findHighest :: [Int] -> (Int, [Int])
findHighest list = 
    (findHighest' list 0, list)
    where 
        findHighest':: [Int] -> Int -> Int
        findHighest' list high = 
            case list of 
                x:xs -> 
                    if x > high then
                        findHighest' xs x
                    else
                        findHighest' xs high
                _ -> 
                    high



setZero :: (Int, [Int]) -> (Int, Int, [Int])
setZero (high, list) = 
    let
        index = fromMaybe 0 $ high `elemIndex` list
    in
        (high, index, zeroNthElement index list)

spread :: (Int, Int, [Int]) -> [Int]
spread (high, index, list) =
    spread' high (index+1) list
    where
        spread':: Int -> Int -> [Int] -> [Int]
        spread' valueToSpread currentIndex list =
            case valueToSpread of
                0 -> list
                _ -> 
                    case getNthElement currentIndex list of
                        Nothing -> 
                            spread' valueToSpread 0 list 
                        Just elem ->
                            spread' (valueToSpread - 1) (currentIndex + 1) (incrementNth currentIndex list)

                            
                


getNthElement :: Int -> [Int] -> Maybe Int
getNthElement index list =
    case splitAt index list of
        (front, element:back) -> Just element
        _ -> Nothing


zeroNthElement :: Int -> [Int] -> [Int]
zeroNthElement index list =
    if index < 0 then 
        list
    else 
        case splitAt index list of
            (front, element:back) -> front ++ 0 : back
            _ -> list

incrementNth :: Int -> [Int] -> [Int]
incrementNth index list =
    if index < 0 then 
        list
    else 
        case splitAt index list of
            (front, element:back) -> front ++ (element + 1) : back
            _ -> list

src/Day7.hs
import System.IO
import Data.List.Split
import Data.List

main = do
    handle <- openFile "input-day-7.txt" ReadMode  
    contents <- hGetContents handle  
    -- print $ fmap (convertToTuple . words) $ lines contents  
    -- print $ fmap words $ lines contents
    print $ getRoot "rmhcw" $ concat $ fmap (convertToParentChild . convertToTuple. words) $ lines contents
    -- print $ convertToTuple ["iqjhw","(119)","->","kwqnczk,","ffulcj"]
    -- print $ concat $ fmap convertToParentChild [
    --           ("ktlj", [])
    --         , ("xhth", [])
    --         , ("ebii", [])
    --         , ("havc", [])
    --         , ("ktlj", [])
    --         , ("padx", ["pbga", "havc", "qoyq"])
    --         , ("fwft", ["ktlj", "cntj", "xhth"])
    --         , ("tknk", ["gyxo", "ebii", "fwft"])
    --     ]
    -- print $ getRoot $ concat $ fmap convertToParentChild [
    --           ("ktlj", [])
    --         , ("xhth", [])
    --         , ("ebii", [])
    --         , ("havc", [])
    --         , ("ktlj", [])
    --         , ("padx", ["pbga", "havc", "qoyq"])
    --         , ("fwft", ["ktlj", "cntj", "xhth"])
    --         , ("tknk", ["gyxo", "ebii", "fwft"])
    --     ]
    hClose handle 

convertToTuple :: [String] -> (String, [String])
convertToTuple list =
    let
        parent = head list
        childList = 
            case splitWhen ((==) "->") list of
                front:back -> fmap removeComma (concat back)
                _ -> []
    in
        (parent, childList)

removeComma :: [Char] -> [Char]
removeComma string =
    delete ',' string

convertToParentChild :: (String, [String]) -> [(String, String)]
convertToParentChild (name, children) = 
    fmap (\childName -> (name, childName)) children 


filterChild :: String -> [(String, String)] -> [(String, String)]
filterChild initial list = 
    filter (\(parent, child) -> child == initial) list  

getRoot :: String -> [(String, String)] -> String
getRoot initial list = 
    getRoot' initial list
    where
        getRoot' childPass list = 
            let
                newList = filter (\(parent, child) -> child == childPass ) list
                (parent, child) = head newList
            in
                case newList of
                    [] -> 
                        childPass
                    _ ->
                        getRoot' parent list 


src/advent1.hs

main:: IO ()
main = print $ result "5228833336355848549915459366737982598312959583817455621545976784792489468198365998232722734876612332352376192813552949814275947575774339529811976644361517795586998319242241614813622734255797569571577699238592667287428166398221572885869416419682687759743978434571821267146514338394624525648338739929479912368172669885577319718389278168766844487948761697438722556857882433224393723131298876252626643517236883999115665656935521675772866516185899317132494716723615493476397115627687887665194781746377341468995954554518252916859227397693885254329628812355612487594445522395853551734567498838382248616137969637971369615443599973588326388792893969924855316437952313492551671545714262784738343517166544197194547173515155927244175447296474282154114951181648317875827525814453758846194548872789943372281952995222779173812444186491115426476188672253249744478946863317915136832199132868917891243591195719354721129116229164688256853628339233919671468781913167415624214152793864585332944468428849171876873433621524242289488135675313544498245498637424139153782925723745249728743885493877792648576673196889949568317234125863369187953788611841388353999875519172896329524346527265231767868839696693328273381772726782949166112932954356923757485139367298699922984925977724972944277991686823219295939734313874834861796179591659174726432357533113896212781566659154939419866797488347448551719481632572231632463575591599696388223344219228325134233238538854289437756331848887242423387542214691157226725179683638967415678697625138177633444765126223885478348951332634398291612134852858683942466178329922655822225426534359191696177633167962839847985826676955417426617126288255366123169174674348417932158291334646767637764323226842771523598562429399935789788215958367362467652444854123951972118358417629679454978687337137675495295768451719631999398617828287671937584998697959425845883145736323818225129311845997214987663433375689621746665629187252511643969315283316269222835744532431378945137649959158495714472963839397214332815241141327714672141875129895"

result :: [Char] -> Int
result a = result' (head a) a
        -- case xs of
        --     [] -> 0
        --     y:ys ->
        --         if x == y then 
        --             read [x] + result xs
        --         else 
        --             0 + result xs


result' :: Char -> [Char] -> Int
result' first input = 
    case input of 
        x:xs -> 
            case xs of
                [] -> 
                    if x == first then read [x]
                    else 0 
                y:_ ->
                    if x == y then 
                        read [x] + result' first xs
                    else 
                        result' first xs
        [] -> 0
    

src/advent3.hs

import Data.List
import Data.Maybe

main = do
    -- print $ createNewList [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25] level (level + 1)
    -- print $ checkLevel 4 1 [ 2,  3,  4,  5,  6,  7,  8,  9]
    -- print $ divideList [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
    -- print $ divideList 2 [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
    -- print $ divideList 1 [ 2,  3,  4,  5,  6,  7,  8,  9]
    -- print $ findEdge 3 (1,[2,3,4,5,6,7,8,9])
    print $ result 4
    -- putStrLn $ "Step for square 4 are: " ++ show square4

result :: Int -> Int
result num = 
    findEdge num $ checkLevel num 1 [ 2,  3,  4,  5,  6,  7,  8,  9]



createNewList :: [Int] -> Int -> Int -> [Int]
createNewList list start final = 
    let
        newList = fmap (+ (8 * start)) list
    in
        if start <= final then
            newList ++ createNewList list (start + 1) final
        else []

checkLevel :: Int -> Int -> [Int] -> (Int, [Int]) 
checkLevel num start list = 
    if num `elem` list then
        (start, list)
    else 
        checkLevel num (start + 1) newList
        where
            newList = createNewList list start (start + 1)

findEdge :: Int ->(Int, [Int]) -> Int
findEdge num (level, list) = 
    sum 
    $ fmap(\edgeList -> 
        if num `elem` edgeList then    
            (fromMaybe 0 (elemIndex num edgeList)) + 1
        else 
            0

    ) 
    $ divideList level list

divideList :: Int -> [Int] -> [[Int]]
divideList level l = 
    [   [(l!!(startIndex))..(l!!(startIndex + space - 1))],
        [(l!!(startIndex + space))..(l!!(startIndex + space*2 - 1))],
        [(l!!(startIndex + space*2))..(l!!(startIndex + space*3 - 1))],
        (
            [(l!!(startIndex + space*3))..(l!!(startIndex + space*4 - (space - level)))]
            -- []
            ++ if startIndex /= 0 then 
                [(head l)..l!!(startIndex - 1)]
            else []
        )

    ] where 
        space = (length l) `div` 4
        startIndex = level - 1
-- 0.[ 1]
-- 1.[ 2,  3,  4,  5,  6,  7,  8,  9]
-- 2.[10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]
-- 3.[26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]
-- -- Find out how to [1, 2, 3, 4] -> [Left, [Down,Left], [Right, Down, Left]]

-- -- input :: [Square]
-- -- input = [Access 1, LeftS 2, TopS 3, RightS 4, RightS 5]

-- 37 36  35  34  33  32 31
-- 38 17  16  15  14  13 30
-- 39 18   5   4   3  12 29
-- 40 19   6   1   2  11 28 53
-- 41 20   7   8   9  10 27 52
-- 42 21  22  23  24  25 26 51
-- 43 44  45  46  47  48 49 50

-- isNumberIn num level list = 
--     if num is not in list then
--         isNumberIn num (level + 1) newList
--     where newList = createNewList list time final
--           createNewList list time = 
--                 if time > final then
--                     list + createNewList (time + 1) final
--                 else []


src/advent4.hs

import System.IO

main:: IO ()
main = do  
    handle <- openFile "advent4.txt" ReadMode  
    contents <- hGetContents handle  
    -- print contents  
    print $ countValid $ fmap checkValid $ convert contents  
    -- print $ checkValid ["test3", "test3", "test"]
    hClose handle 


convert :: String -> [[String]]
convert input = 
    fmap words $ lines input

-- checkSentence :: String -> Int
-- checkSentence sentence = 
--     countDuplicate $ word setence

checkValid :: [String] -> Bool
checkValid words =
    checkValid' words []
    
checkValid' :: [String] -> [String] -> Bool
checkValid' words uniqueWords =
    case words of
        x:xs ->
            if x `elem` uniqueWords then
                False
            else 
                checkValid' xs (x:uniqueWords)
        _ -> 
            True

countValid :: [Bool] -> Int
countValid boolList = 
    countValid' 0 boolList

countValid' :: Int -> [Bool] -> Int
countValid' sum boolList = 
    case boolList of
        x:xs ->
            if x == True then
                countValid' (sum + 1) xs
            else 
                countValid' sum xs
        _ ->
            sum



src/advent5.hs

import System.IO

main = do
    handle <- openFile "advent5.txt" ReadMode  
    contents <- hGetContents handle  
    -- print contents  
    print $ countStep $ convert contents  
    -- print $ checkValid ["test3", "test3", "test"]
    hClose handle 


    -- print $ countStep' (0, 0, [0, 3, 0, 1, -3])
    -- print $ getNthElement 5 [5, 3, 0, 1, -3]
    -- print $ countStep [0, 3, 0, 1, -3]

-- countStep :: [Int] -> Int
-- countStep input = 
--     let
--         (_, step, _) = countStep' (0, 0, input)
--     in
--         step
convert :: String -> [Int]
convert input = 
    fmap read $ lines input

countStep :: [Int] -> (Int, Int, [Int])
countStep inputList = countStep' (0, 0, inputList)

countStep' :: (Int, Int, [Int]) -> (Int, Int, [Int]) 
countStep' (index, step, input) = 
    case getNthElement index input of
        Nothing ->
            (index, step, input)
        Just elem ->
            countStep' (index + elem, step + 1, changeNthElement index input)


getNthElement :: Int -> [Int] -> Maybe Int
getNthElement index list =
    case splitAt index list of
        (front, element:back) -> Just element
        _ -> Nothing
    
changeNthElement :: Int -> [Int] -> [Int]
changeNthElement index list
    | index < 0   = list
    | otherwise = case splitAt index list of
                    (front, element:back) -> front ++ (element + 1) : back
                    _ -> list

src/Day3.1.hs

import Data.List
import Data.Maybe

main = do
    -- print $ createNewList [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25] level (level + 1)
    -- print $ checkLevel 4 1 [ 2,  3,  4,  5,  6,  7,  8,  9]
    -- print $ divideList [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
    -- print $ divideList 2 [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
    -- print $ divideList 1 [ 2,  3,  4,  5,  6,  7,  8,  9]
    -- print $ findEdge 3 (1,[2,3,4,5,6,7,8,9])
    -- print $ result 49
    -- print $ findEdge 9 $ checkLevel 9 1 (2, 9)
    print $ final 368078
    -- print $ checkLevel 49 1 (2, 9) 
    -- putStrLn $ "Step for square 4 are: " ++ show square4

-- result :: Int -> Int
result num = 
    -- findEdge num $ checkLevel num 1 (2, 9)
    let
        (level, (start, end)) = checkLevel num 1 (2, 9)
    in
        divideList level [start..end]

final num = 
    findEdge num $ checkLevel num 1 (2, 9)


-- createNewFirstEnd :: (Int, Int) -> Int -> Int -> (Int, Int)
-- createNewFirstEnd (first, end) start final = 
--     let
--         newList = fmap (+ (8 * start)) list
--     in
--         if start <= final then
--             -- newList ++ createNewFirstEnd list (start + 1) final
--             (,) first 
--         else []

checkLevel :: Int -> Int -> (Int, Int) -> (Int, (Int, Int)) 
checkLevel num level (first, end) = 
    if first <= num && num <= end then
        (level, (first, end))
    else 
        checkLevel num (level + 1) newFirstEnd
        where
            newFirstEnd = (first + 8*level, end + 8*(level+1))
        -- where
        --     newFirstEnd = createNewFirstEnd list defaultLevel (defaultLevel + 1)

findEdge :: Int -> (Int, (Int, Int)) -> Int
findEdge num (level, (start, end)) = 
    sum 
    -- $ fmap(\index ->
    --     index * (level - 1)     
    -- )
    $ fmap(\edgeList -> 
        if num `elem` edgeList then    
            let index = (fromMaybe 0 (elemIndex num edgeList)) 
            in  if index >= ((length edgeList) - (level - 1)) then
                    (length edgeList - index) + (level)
                    -- (length edgeList - index)
                    -- (length edgeList - index)
                else 
                    index + (level)
                    -- index
        else 
            0

    ) 
    $ divideList level [start..end]

divideList :: Int -> [Int] -> [[Int]]
divideList level l = 
    [   [(l!!(startIndex))..(l!!(startIndex + space - 1))],
        [(l!!(startIndex + space))..(l!!(startIndex + space*2 - 1))],
        [(l!!(startIndex + space*2))..(l!!(startIndex + space*3 - 1))],
        (
            [(l!!(startIndex + space*3))..(l!!(startIndex + space*4 - (space - level)))]
            -- []
            ++ if startIndex /= 0 then 
                [(head l)..l!!(startIndex - 1)]
            else []
        )

    ] where 
        space = (length l) `div` 4
        startIndex = level - 1
-- 0.[ 1]
-- 1.[ 2,  3,  4,  5,  6,  7,  8,  9]
-- 2.[10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]
-- 3.[26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]
-- -- Find out how to [1, 2, 3, 4] -> [Left, [Down,Left], [Right, Down, Left]]

-- -- input :: [Square]
-- -- input = [Access 1, LeftS 2, TopS 3, RightS 4, RightS 5]

-- 37 36  35  34  33  32 31
-- 38 17  16  15  14  13 30
-- 39 18   5   4   3  12 29
-- 40 19   6   1   2  11 28 53
-- 41 20   7   8   9  10 27 52
-- 42 21  22  23  24  25 26 51
-- 43 44  45  46  47  48 49 50

-- isNumberIn num level list = 
--     if num is not in list then
--         isNumberIn num (level + 1) newList
--     where newList = createNewList list time final
--           createNewList list time = 
--                 if time > final then
--                     list + createNewList (time + 1) final
--                 else []
